server/profiler.py

```diff
@@ -0,0 +1,93 @@
+import asyncio
+import cProfile
+import logging
+
+from server.config import config
+from server.decorators import with_logger
+from server.timing import at_interval
+
+
+@with_logger
+class Profiler:
+    def __init__(
+        self, interval, player_service, duration=2, max_count=300, outfile="profile.txt"
+    ):
+        self.profiler = cProfile.Profile()
+        self.interval = interval
+        self.duration = duration
+        self.current_count = 0
+        self.max_count = max_count
+
+        self._player_service = player_service
+        self._outfile = outfile
+
+        self._running = False
+        self._task = None
+        self._loop = asyncio.get_running_loop()
+
+    def start(self):
+        self._running = True
+        if self._task is None:
+            self._task = asyncio.create_task(Profiler._next_run(self))
+
+    async def _next_run(self):
+        await asyncio.sleep(self.interval)
+
+        if not self._running:
+            return
+        self.current_count += 1
```

Comment:
Right, I did that intentionally thinking that if the server is under high load for a long time, you don't want the profiling coroutine to keep running until the load drops (if it ever does - right now the player count is rarely dropping below 500 I think, not below 750 on weekends).
So this is rather the count of "attempts" than "runs".
We could reflect that in the variable names?

Reply:
The coroutine running isn't going to cause much load, calling `profile.enable()` might. Although I think it's actually not that bad in python since the profiler is implemented in C and python is already so slow. Might be worth bumping up the "high load" threshold since even at 2000 players we're only at about 20% cpu usage. I had it lower before because when I was awaiting broadcasts it was maxing out at about 1500.

Reply:
I think only increment if the profiler actually runs. When I get the profile file I want it to contain the full amount of data regardless of when the profiler was started.